storages/backends/gcloud.py

```diff
@@ -0,0 +1,233 @@
+from tempfile import SpooledTemporaryFile
+
+from django.core.exceptions import ImproperlyConfigured
+from django.core.files.base import File
+from django.core.files.storage import Storage
+from django.utils.deconstruct import deconstructible
+from django.utils.encoding import force_bytes, force_text, smart_str
+from django.utils import timezone
+from storages.utils import clean_name, safe_join, setting
+
+try:
+    from google.cloud.storage.client import Client
+    from google.cloud.storage.blob import Blob
+    from google.cloud.exceptions import NotFound
+except ImportError:
+    raise ImproperlyConfigured("Could not load Google Cloud Storage bindings.\n"
+                               "See https://github.com/GoogleCloudPlatform/gcloud-python")
+
+
+class GoogleCloudFile(File):
+    def __init__(self, name, mode, storage):
+        self.name = name
+        self._mode = mode
+        self._storage = storage
+        self.blob = storage.bucket.get_blob(name)
+        if not self.blob and 'w' in mode:
+            self.blob = Blob(self.name, storage.bucket)
+        self._file = None
+        self._is_dirty = False
+
+    @property
+    def size(self):
+        return self.blob.size
+
+    def _get_file(self):
+        if self._file is None:
+            self._file = SpooledTemporaryFile(
+                max_size=self._storage.max_memory_size,
+                suffix=".GSStorageFile",
+                dir=setting("FILE_UPLOAD_TEMP_DIR", None)
+            )
+            if 'r' in self._mode:
+                self._is_dirty = False
+                self.blob.download_to_file(self._file)
+                self._file.seek(0)
+        return self._file
+
+    def _set_file(self, value):
+        self._file = value
+
+    file = property(_get_file, _set_file)
+
+    def read(self, num_bytes=None):
+        if 'r' not in self._mode:
+            raise AttributeError("File was not opened in read mode.")
+        if num_bytes is not None:
+            return super(GoogleCloudFile, self).read(num_bytes)
+        else:
+            return super(GoogleCloudFile, self).read()
+
+    def write(self, content):
+        if 'w' not in self._mode:
+            raise AttributeError("File was not opened in write mode.")
+        self._is_dirty = True
+        return super(GoogleCloudFile, self).write(force_bytes(content))
+
+    def close(self):
+        if self._file is not None:
+            if self._is_dirty:
+                self.file.seek(0)
+                self.blob.upload_from_file(self.file)
+            self._file.close()
+            self._file = None
+
+
+@deconstructible
+class GoogleCloudStorage(Storage):
+    project_id = setting('GS_PROJECT_ID', None)
+    credentials = setting('GS_CREDENTIALS', None)
+    bucket_name = setting('GS_BUCKET_NAME', None)
+    auto_create_bucket = setting('GS_AUTO_CREATE_BUCKET', False)
+    default_acl = setting('GS_DEFAULT_ACL', 'public-read')
```

Comment:
This is only used in the case where we create the bucket (`auto_create_bucket = True`). In that case we need a default because there are no default object ACLs we can use.